{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! pip install mlflow -q"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import required libraries, data and results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import anndata\n",
    "import xgboost\n",
    "from pathlib import Path\n",
    "\n",
    "base_dataset_path = \"/data/perturbation_bining_colon_epothelial_take_7/base_dataset.h5ad\"\n",
    "control_data_path = \"/data/perturbation_bining_colon_epothelial_take_7/perturbation_experiment_JUND_level_1.0.embeddings.npz\"\n",
    "knockout_data_path = \"/data/perturbation_bining_colon_epothelial_take_7/perturbation_experiment_JUND_level_0.embeddings.npz\"\n",
    "\n",
    "# load the datasets\n",
    "if not \"adata\" in locals():\n",
    "    adata = anndata.read_h5ad(base_dataset_path)\n",
    "    control_embedding = np.load(control_data_path)[\"emb\"]\n",
    "    ko_embedding = np.load(knockout_data_path)[\"emb\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The reaults from the scGPT capsule: embeddings CELL x Vector (97788, 512) matrix\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[ 0.40102956,  1.1505572 , -0.846087  , ..., -0.22407077,\n",
       "        -0.617475  ,  0.15734558],\n",
       "       [ 0.41551918,  1.0006112 , -0.5593975 , ..., -0.22076431,\n",
       "        -0.6000564 ,  0.09469572],\n",
       "       [ 0.40629452,  0.94172704, -0.50550914, ..., -0.25236434,\n",
       "        -0.48028022, -0.013081  ],\n",
       "       ...,\n",
       "       [ 1.0662296 ,  1.412075  ,  0.11929194, ..., -0.5681815 ,\n",
       "        -0.5062771 , -0.19895682],\n",
       "       [ 0.62741625,  1.1626298 , -0.16092153, ..., -0.6266747 ,\n",
       "         0.01310916, -0.31598997],\n",
       "       [ 0.9090164 ,  1.4771746 , -0.1576058 , ..., -0.70694625,\n",
       "        -0.3748046 , -0.49906144]], dtype=float32)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The reaults from the scGPT capsule: embeddings CELL x Vector {ko_embedding.shape} matrix\")\n",
    "ko_embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Single-cell transcriptomics data shape: cells: 97,788, genes: 27,289\n",
      "control ('WT')  embedding shape:        cells: 97,788 features: 512\n",
      "KnockOut ('KO') embedding shape:        cells: 97,788 features: 512\n"
     ]
    }
   ],
   "source": [
    "print(f\"Single-cell transcriptomics data shape: cells: {adata.shape[0]:,}, genes: {adata.shape[1]:,}\")\n",
    "print(f\"control ('WT')  embedding shape:        cells: {control_embedding.shape[0]:,} features: {control_embedding.shape[1]}\")\n",
    "print(f\"KnockOut ('KO') embedding shape:        cells: {ko_embedding.shape[0]:,} features: {ko_embedding.shape[1]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The main metdata of the cell transcriptomics dataset\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>N175041_N1-TAGAGCTCAGGAATGC</th>\n",
       "      <td>goblet cell</td>\n",
       "      <td>seventh decade human stage</td>\n",
       "      <td>left colon</td>\n",
       "      <td>Crohn disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N178961_L-AGGTTGTCACAGTGAG</th>\n",
       "      <td>enterocyte</td>\n",
       "      <td>seventh decade human stage</td>\n",
       "      <td>lamina propria of mucosa of colon</td>\n",
       "      <td>Crohn disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H197396_N2-GTACTCCCACGGTAAG</th>\n",
       "      <td>goblet cell</td>\n",
       "      <td>eighth decade human stage</td>\n",
       "      <td>colon</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N128624_E-GAAGGGTGTATCGCTA</th>\n",
       "      <td>enterocyte</td>\n",
       "      <td>fifth decade human stage</td>\n",
       "      <td>colonic epithelium</td>\n",
       "      <td>Crohn disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H197396_N3-GATCGATGTTGCGTTA</th>\n",
       "      <td>enterocyte</td>\n",
       "      <td>eighth decade human stage</td>\n",
       "      <td>colon</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N15_Epi_B-TAGGGACTGAGGTG</th>\n",
       "      <td>enterocyte</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>right colon</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N130084_E-TCATTTGTCAACCCGG</th>\n",
       "      <td>enterocyte</td>\n",
       "      <td>fifth decade human stage</td>\n",
       "      <td>colonic epithelium</td>\n",
       "      <td>Crohn disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I114902_N-CACACCTCAAGTTGTC</th>\n",
       "      <td>goblet cell</td>\n",
       "      <td>fourth decade human stage</td>\n",
       "      <td>colon</td>\n",
       "      <td>Crohn disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H197396_N3-AGGGATGCAACAACCT</th>\n",
       "      <td>stem cell</td>\n",
       "      <td>eighth decade human stage</td>\n",
       "      <td>colon</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N10_Epi_A-TGCGAAACATTCCT</th>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>right colon</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97788 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell_type           development_stage  \\\n",
       "cell_id                                                                    \n",
       "N175041_N1-TAGAGCTCAGGAATGC      goblet cell  seventh decade human stage   \n",
       "N178961_L-AGGTTGTCACAGTGAG        enterocyte  seventh decade human stage   \n",
       "H197396_N2-GTACTCCCACGGTAAG      goblet cell   eighth decade human stage   \n",
       "N128624_E-GAAGGGTGTATCGCTA        enterocyte    fifth decade human stage   \n",
       "H197396_N3-GATCGATGTTGCGTTA       enterocyte   eighth decade human stage   \n",
       "...                                      ...                         ...   \n",
       "N15_Epi_B-TAGGGACTGAGGTG          enterocyte           human adult stage   \n",
       "N130084_E-TCATTTGTCAACCCGG        enterocyte    fifth decade human stage   \n",
       "I114902_N-CACACCTCAAGTTGTC       goblet cell   fourth decade human stage   \n",
       "H197396_N3-AGGGATGCAACAACCT        stem cell   eighth decade human stage   \n",
       "N10_Epi_A-TGCGAAACATTCCT     epithelial cell           human adult stage   \n",
       "\n",
       "                                                        tissue        disease  \n",
       "cell_id                                                                        \n",
       "N175041_N1-TAGAGCTCAGGAATGC                         left colon  Crohn disease  \n",
       "N178961_L-AGGTTGTCACAGTGAG   lamina propria of mucosa of colon  Crohn disease  \n",
       "H197396_N2-GTACTCCCACGGTAAG                              colon         normal  \n",
       "N128624_E-GAAGGGTGTATCGCTA                  colonic epithelium  Crohn disease  \n",
       "H197396_N3-GATCGATGTTGCGTTA                              colon         normal  \n",
       "...                                                        ...            ...  \n",
       "N15_Epi_B-TAGGGACTGAGGTG                           right colon         normal  \n",
       "N130084_E-TCATTTGTCAACCCGG                  colonic epithelium  Crohn disease  \n",
       "I114902_N-CACACCTCAAGTTGTC                               colon  Crohn disease  \n",
       "H197396_N3-AGGGATGCAACAACCT                              colon         normal  \n",
       "N10_Epi_A-TGCGAAACATTCCT                           right colon         normal  \n",
       "\n",
       "[97788 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explore the datasets\n",
    "print(\"The main metdata of the cell transcriptomics dataset\")\n",
    "adata.obs[[\"cell_type\", \"development_stage\", \"tissue\", \"disease\"]].sample(frac=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Train a model to seperate the Crohn's and Normal cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: black;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"â–¸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"â–¾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 1ex;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>XGBClassifier(base_score=None, booster=None, callbacks=None,\n",
       "              colsample_bylevel=None, colsample_bynode=None,\n",
       "              colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "              enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "              gamma=None, grow_policy=None, importance_type=None,\n",
       "              interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "              max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "              max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "              min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "              multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "              num_parallel_tree=None, random_state=None, ...)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;XGBClassifier<span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>XGBClassifier(base_score=None, booster=None, callbacks=None,\n",
       "              colsample_bylevel=None, colsample_bynode=None,\n",
       "              colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "              enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "              gamma=None, grow_policy=None, importance_type=None,\n",
       "              interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "              max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "              max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "              min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "              multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "              num_parallel_tree=None, random_state=None, ...)</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "XGBClassifier(base_score=None, booster=None, callbacks=None,\n",
       "              colsample_bylevel=None, colsample_bynode=None,\n",
       "              colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "              enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "              gamma=None, grow_policy=None, importance_type=None,\n",
       "              interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "              max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "              max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "              min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "              multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "              num_parallel_tree=None, random_state=None, ...)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# train a classifier on the control_embedding_data\n",
    "trained_model_path = \"xgboost_model.pkl\"\n",
    "\n",
    "if Path(trained_model_path).exists():\n",
    "    # load the trained model\n",
    "    with open(trained_model_path, \"rb\") as f:\n",
    "        model = pickle.load(f)\n",
    "\n",
    "else:\n",
    "    # convert the disease data to numerical:\n",
    "    disease_categories = pd.Categorical(adata.obs[\"disease\"]).codes\n",
    "    model = xgboost.XGBClassifier() # initialize the model\n",
    "    model.fit(control_embedding, disease_categories) # train the model\n",
    "    \n",
    "    # save the model\n",
    "    with open(trained_model_path, \"wb\") as f:\n",
    "        pickle.dump(model, f)\n",
    "\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use the model and the results from scGPT to estimate the effect of KO on the Crohn's state"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H197396_N1-CGTCACTGTTAAGAAC</th>\n",
       "      <td>colon</td>\n",
       "      <td>enterocyte</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N15_Epi_A-GTAACGTGGTATGC</th>\n",
       "      <td>right colon</td>\n",
       "      <td>stem cell</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N10_Epi_B-AGGCAACTATGTGC</th>\n",
       "      <td>right colon</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H197396_N4-CAGAGAGGTGGGTATG</th>\n",
       "      <td>colon</td>\n",
       "      <td>enterocyte</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N114902_N-CAAGATCGTTCGTGAT</th>\n",
       "      <td>colon</td>\n",
       "      <td>goblet cell</td>\n",
       "      <td>Crohn disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N13_Epi_A-CGTGTAGACGCCTT</th>\n",
       "      <td>right colon</td>\n",
       "      <td>goblet cell</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I175041_N2-CACACTCCAGCTGCTG</th>\n",
       "      <td>sigmoid colon</td>\n",
       "      <td>enterocyte</td>\n",
       "      <td>Crohn disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N51_Epi_B-TCGGGACAGAGCTTCT</th>\n",
       "      <td>caecum</td>\n",
       "      <td>goblet cell</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I175041_N1-CTAACTTGTATCTGCA</th>\n",
       "      <td>sigmoid colon</td>\n",
       "      <td>enterocyte</td>\n",
       "      <td>Crohn disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N18_Epi_A-ATTGCTTGATCAGC</th>\n",
       "      <td>transverse colon</td>\n",
       "      <td>enterocyte</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       tissue        cell_type        disease\n",
       "cell_id                                                                      \n",
       "H197396_N1-CGTCACTGTTAAGAAC             colon       enterocyte         normal\n",
       "N15_Epi_A-GTAACGTGGTATGC          right colon        stem cell         normal\n",
       "N10_Epi_B-AGGCAACTATGTGC          right colon  epithelial cell         normal\n",
       "H197396_N4-CAGAGAGGTGGGTATG             colon       enterocyte         normal\n",
       "N114902_N-CAAGATCGTTCGTGAT              colon      goblet cell  Crohn disease\n",
       "N13_Epi_A-CGTGTAGACGCCTT          right colon      goblet cell         normal\n",
       "I175041_N2-CACACTCCAGCTGCTG     sigmoid colon       enterocyte  Crohn disease\n",
       "N51_Epi_B-TCGGGACAGAGCTTCT             caecum      goblet cell         normal\n",
       "I175041_N1-CTAACTTGTATCTGCA     sigmoid colon       enterocyte  Crohn disease\n",
       "N18_Epi_A-ATTGCTTGATCAGC     transverse colon       enterocyte         normal"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the cells' metadata\n",
    "result = adata.obs[[\"tissue\",\"cell_type\", \"disease\"]].copy()\n",
    "result.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_65443_row0_col3 {\n",
       "  background-color: #4257c9;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_65443_row1_col3 {\n",
       "  background-color: #f7a889;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_65443_row2_col3 {\n",
       "  background-color: #b40426;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_65443_row3_col3, #T_65443_row6_col3 {\n",
       "  background-color: #5d7ce6;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_65443_row4_col3 {\n",
       "  background-color: #4055c8;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_65443_row5_col3 {\n",
       "  background-color: #3b4cc0;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_65443_row7_col3 {\n",
       "  background-color: #9dbdff;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_65443_row8_col3 {\n",
       "  background-color: #536edd;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_65443_row9_col3 {\n",
       "  background-color: #8badfd;\n",
       "  color: #000000;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_65443\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_65443_level0_col0\" class=\"col_heading level0 col0\" >tissue</th>\n",
       "      <th id=\"T_65443_level0_col1\" class=\"col_heading level0 col1\" >cell_type</th>\n",
       "      <th id=\"T_65443_level0_col2\" class=\"col_heading level0 col2\" >disease</th>\n",
       "      <th id=\"T_65443_level0_col3\" class=\"col_heading level0 col3\" >control_prediction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >cell_id</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row0\" class=\"row_heading level0 row0\" >I175041_N1-ACCAGTACACTAGTAC</th>\n",
       "      <td id=\"T_65443_row0_col0\" class=\"data row0 col0\" >sigmoid colon</td>\n",
       "      <td id=\"T_65443_row0_col1\" class=\"data row0 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row0_col2\" class=\"data row0 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_65443_row0_col3\" class=\"data row0 col3\" >0.046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row1\" class=\"row_heading level0 row1\" >H197396_N1-TTGGCAAAGATATGGT</th>\n",
       "      <td id=\"T_65443_row1_col0\" class=\"data row1 col0\" >colon</td>\n",
       "      <td id=\"T_65443_row1_col1\" class=\"data row1 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row1_col2\" class=\"data row1 col2\" >normal</td>\n",
       "      <td id=\"T_65443_row1_col3\" class=\"data row1 col3\" >0.660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row2\" class=\"row_heading level0 row2\" >N20_Epi_B-AGATTCCTCCTTTA</th>\n",
       "      <td id=\"T_65443_row2_col0\" class=\"data row2 col0\" >colon</td>\n",
       "      <td id=\"T_65443_row2_col1\" class=\"data row2 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row2_col2\" class=\"data row2 col2\" >normal</td>\n",
       "      <td id=\"T_65443_row2_col3\" class=\"data row2 col3\" >0.919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row3\" class=\"row_heading level0 row3\" >N128400_L-ACTACGATCGTTCTAT</th>\n",
       "      <td id=\"T_65443_row3_col0\" class=\"data row3 col0\" >lamina propria of mucosa of colon</td>\n",
       "      <td id=\"T_65443_row3_col1\" class=\"data row3 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row3_col2\" class=\"data row3 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_65443_row3_col3\" class=\"data row3 col3\" >0.122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row4\" class=\"row_heading level0 row4\" >N178961_L-CTAACTTCACCCTAGG</th>\n",
       "      <td id=\"T_65443_row4_col0\" class=\"data row4 col0\" >lamina propria of mucosa of colon</td>\n",
       "      <td id=\"T_65443_row4_col1\" class=\"data row4 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row4_col2\" class=\"data row4 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_65443_row4_col3\" class=\"data row4 col3\" >0.041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row5\" class=\"row_heading level0 row5\" >H180844_N2-TGACAACTCACCCTCA</th>\n",
       "      <td id=\"T_65443_row5_col0\" class=\"data row5 col0\" >colon</td>\n",
       "      <td id=\"T_65443_row5_col1\" class=\"data row5 col1\" >epithelial cell</td>\n",
       "      <td id=\"T_65443_row5_col2\" class=\"data row5 col2\" >normal</td>\n",
       "      <td id=\"T_65443_row5_col3\" class=\"data row5 col3\" >0.022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row6\" class=\"row_heading level0 row6\" >N128624_E-TATTCCACAAGTAGTA</th>\n",
       "      <td id=\"T_65443_row6_col0\" class=\"data row6 col0\" >colonic epithelium</td>\n",
       "      <td id=\"T_65443_row6_col1\" class=\"data row6 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row6_col2\" class=\"data row6 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_65443_row6_col3\" class=\"data row6 col3\" >0.121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row7\" class=\"row_heading level0 row7\" >N130084_E-GGAGGTACAGGTTCCG</th>\n",
       "      <td id=\"T_65443_row7_col0\" class=\"data row7 col0\" >colonic epithelium</td>\n",
       "      <td id=\"T_65443_row7_col1\" class=\"data row7 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row7_col2\" class=\"data row7 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_65443_row7_col3\" class=\"data row7 col3\" >0.288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row8\" class=\"row_heading level0 row8\" >N51_Epi_A-CGGTTAAGTTCATGGT</th>\n",
       "      <td id=\"T_65443_row8_col0\" class=\"data row8 col0\" >caecum</td>\n",
       "      <td id=\"T_65443_row8_col1\" class=\"data row8 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row8_col2\" class=\"data row8 col2\" >normal</td>\n",
       "      <td id=\"T_65443_row8_col3\" class=\"data row8 col3\" >0.095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_65443_level0_row9\" class=\"row_heading level0 row9\" >N175041_N2-CCTAGCTGTACCGGCT</th>\n",
       "      <td id=\"T_65443_row9_col0\" class=\"data row9 col0\" >left colon</td>\n",
       "      <td id=\"T_65443_row9_col1\" class=\"data row9 col1\" >enterocyte</td>\n",
       "      <td id=\"T_65443_row9_col2\" class=\"data row9 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_65443_row9_col3\" class=\"data row9 col3\" >0.240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7f21e5760490>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the probability of chron's disease for each cell, based on the model:\n",
    "result[\"control_prediction\"] = model.predict_proba(control_embedding)[:, 1]\n",
    "\n",
    "# visualize:\n",
    "(\n",
    "    result.sample(10, random_state=23)\n",
    "    .style.format(\"{:.3f}\", subset=[\"control_prediction\"])\n",
    "    .background_gradient(subset=[\"control_prediction\"], cmap=\"coolwarm\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_2593e_row0_col3 {\n",
       "  background-color: #4257c9;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row0_col4 {\n",
       "  background-color: #5977e3;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row1_col3 {\n",
       "  background-color: #f7a889;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_2593e_row1_col4 {\n",
       "  background-color: #c73635;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row2_col3, #T_2593e_row2_col4 {\n",
       "  background-color: #b40426;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row3_col3, #T_2593e_row6_col3 {\n",
       "  background-color: #5d7ce6;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row3_col4 {\n",
       "  background-color: #dddcdc;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_2593e_row4_col3, #T_2593e_row4_col4 {\n",
       "  background-color: #4055c8;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row5_col3, #T_2593e_row5_col4 {\n",
       "  background-color: #3b4cc0;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row6_col4 {\n",
       "  background-color: #3e51c5;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row7_col3 {\n",
       "  background-color: #9dbdff;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_2593e_row7_col4 {\n",
       "  background-color: #7396f5;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row8_col3 {\n",
       "  background-color: #536edd;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row8_col4 {\n",
       "  background-color: #7295f4;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_2593e_row9_col3 {\n",
       "  background-color: #8badfd;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_2593e_row9_col4 {\n",
       "  background-color: #c43032;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_2593e\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_2593e_level0_col0\" class=\"col_heading level0 col0\" >tissue</th>\n",
       "      <th id=\"T_2593e_level0_col1\" class=\"col_heading level0 col1\" >cell_type</th>\n",
       "      <th id=\"T_2593e_level0_col2\" class=\"col_heading level0 col2\" >disease</th>\n",
       "      <th id=\"T_2593e_level0_col3\" class=\"col_heading level0 col3\" >control_prediction</th>\n",
       "      <th id=\"T_2593e_level0_col4\" class=\"col_heading level0 col4\" >ko_prediction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >cell_id</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "      <th class=\"blank col4\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row0\" class=\"row_heading level0 row0\" >I175041_N1-ACCAGTACACTAGTAC</th>\n",
       "      <td id=\"T_2593e_row0_col0\" class=\"data row0 col0\" >sigmoid colon</td>\n",
       "      <td id=\"T_2593e_row0_col1\" class=\"data row0 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row0_col2\" class=\"data row0 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_2593e_row0_col3\" class=\"data row0 col3\" >0.046</td>\n",
       "      <td id=\"T_2593e_row0_col4\" class=\"data row0 col4\" >0.099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row1\" class=\"row_heading level0 row1\" >H197396_N1-TTGGCAAAGATATGGT</th>\n",
       "      <td id=\"T_2593e_row1_col0\" class=\"data row1 col0\" >colon</td>\n",
       "      <td id=\"T_2593e_row1_col1\" class=\"data row1 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row1_col2\" class=\"data row1 col2\" >normal</td>\n",
       "      <td id=\"T_2593e_row1_col3\" class=\"data row1 col3\" >0.660</td>\n",
       "      <td id=\"T_2593e_row1_col4\" class=\"data row1 col4\" >0.899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row2\" class=\"row_heading level0 row2\" >N20_Epi_B-AGATTCCTCCTTTA</th>\n",
       "      <td id=\"T_2593e_row2_col0\" class=\"data row2 col0\" >colon</td>\n",
       "      <td id=\"T_2593e_row2_col1\" class=\"data row2 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row2_col2\" class=\"data row2 col2\" >normal</td>\n",
       "      <td id=\"T_2593e_row2_col3\" class=\"data row2 col3\" >0.919</td>\n",
       "      <td id=\"T_2593e_row2_col4\" class=\"data row2 col4\" >0.949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row3\" class=\"row_heading level0 row3\" >N128400_L-ACTACGATCGTTCTAT</th>\n",
       "      <td id=\"T_2593e_row3_col0\" class=\"data row3 col0\" >lamina propria of mucosa of colon</td>\n",
       "      <td id=\"T_2593e_row3_col1\" class=\"data row3 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row3_col2\" class=\"data row3 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_2593e_row3_col3\" class=\"data row3 col3\" >0.122</td>\n",
       "      <td id=\"T_2593e_row3_col4\" class=\"data row3 col4\" >0.478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row4\" class=\"row_heading level0 row4\" >N178961_L-CTAACTTCACCCTAGG</th>\n",
       "      <td id=\"T_2593e_row4_col0\" class=\"data row4 col0\" >lamina propria of mucosa of colon</td>\n",
       "      <td id=\"T_2593e_row4_col1\" class=\"data row4 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row4_col2\" class=\"data row4 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_2593e_row4_col3\" class=\"data row4 col3\" >0.041</td>\n",
       "      <td id=\"T_2593e_row4_col4\" class=\"data row4 col4\" >0.025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row5\" class=\"row_heading level0 row5\" >H180844_N2-TGACAACTCACCCTCA</th>\n",
       "      <td id=\"T_2593e_row5_col0\" class=\"data row5 col0\" >colon</td>\n",
       "      <td id=\"T_2593e_row5_col1\" class=\"data row5 col1\" >epithelial cell</td>\n",
       "      <td id=\"T_2593e_row5_col2\" class=\"data row5 col2\" >normal</td>\n",
       "      <td id=\"T_2593e_row5_col3\" class=\"data row5 col3\" >0.022</td>\n",
       "      <td id=\"T_2593e_row5_col4\" class=\"data row5 col4\" >0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row6\" class=\"row_heading level0 row6\" >N128624_E-TATTCCACAAGTAGTA</th>\n",
       "      <td id=\"T_2593e_row6_col0\" class=\"data row6 col0\" >colonic epithelium</td>\n",
       "      <td id=\"T_2593e_row6_col1\" class=\"data row6 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row6_col2\" class=\"data row6 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_2593e_row6_col3\" class=\"data row6 col3\" >0.121</td>\n",
       "      <td id=\"T_2593e_row6_col4\" class=\"data row6 col4\" >0.018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row7\" class=\"row_heading level0 row7\" >N130084_E-GGAGGTACAGGTTCCG</th>\n",
       "      <td id=\"T_2593e_row7_col0\" class=\"data row7 col0\" >colonic epithelium</td>\n",
       "      <td id=\"T_2593e_row7_col1\" class=\"data row7 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row7_col2\" class=\"data row7 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_2593e_row7_col3\" class=\"data row7 col3\" >0.288</td>\n",
       "      <td id=\"T_2593e_row7_col4\" class=\"data row7 col4\" >0.171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row8\" class=\"row_heading level0 row8\" >N51_Epi_A-CGGTTAAGTTCATGGT</th>\n",
       "      <td id=\"T_2593e_row8_col0\" class=\"data row8 col0\" >caecum</td>\n",
       "      <td id=\"T_2593e_row8_col1\" class=\"data row8 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row8_col2\" class=\"data row8 col2\" >normal</td>\n",
       "      <td id=\"T_2593e_row8_col3\" class=\"data row8 col3\" >0.095</td>\n",
       "      <td id=\"T_2593e_row8_col4\" class=\"data row8 col4\" >0.167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_2593e_level0_row9\" class=\"row_heading level0 row9\" >N175041_N2-CCTAGCTGTACCGGCT</th>\n",
       "      <td id=\"T_2593e_row9_col0\" class=\"data row9 col0\" >left colon</td>\n",
       "      <td id=\"T_2593e_row9_col1\" class=\"data row9 col1\" >enterocyte</td>\n",
       "      <td id=\"T_2593e_row9_col2\" class=\"data row9 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_2593e_row9_col3\" class=\"data row9 col3\" >0.240</td>\n",
       "      <td id=\"T_2593e_row9_col4\" class=\"data row9 col4\" >0.908</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7f21b5b417e0>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result[\"ko_prediction\"] = model.predict_proba(ko_embedding)[:, 1]\n",
    "\n",
    "# visualize:\n",
    "(\n",
    "    result.sample(10, random_state=23)\n",
    "    .style.format(\"{:.3f}\", subset=[\"control_prediction\",'ko_prediction'])\n",
    "    .background_gradient(subset=[\"control_prediction\",'ko_prediction'], cmap=\"coolwarm\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## calculate the 'delta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_11054_row0_col3 {\n",
       "  background-color: #4257c9;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row0_col4 {\n",
       "  background-color: #5977e3;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row0_col5 {\n",
       "  background-color: #f08b6e;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row1_col3 {\n",
       "  background-color: #f7a889;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_11054_row1_col4 {\n",
       "  background-color: #c73635;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row1_col5 {\n",
       "  background-color: #e8d6cc;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_11054_row2_col3, #T_11054_row2_col4, #T_11054_row7_col5 {\n",
       "  background-color: #b40426;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row2_col5 {\n",
       "  background-color: #ec7f63;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row3_col3, #T_11054_row6_col3 {\n",
       "  background-color: #5d7ce6;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row3_col4 {\n",
       "  background-color: #dddcdc;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_11054_row3_col5 {\n",
       "  background-color: #bfd3f6;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_11054_row4_col3, #T_11054_row4_col4 {\n",
       "  background-color: #4055c8;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row4_col5 {\n",
       "  background-color: #df634e;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row5_col3, #T_11054_row5_col4, #T_11054_row9_col5 {\n",
       "  background-color: #3b4cc0;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row5_col5 {\n",
       "  background-color: #de614d;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row6_col4 {\n",
       "  background-color: #3e51c5;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row6_col5 {\n",
       "  background-color: #ba162b;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row7_col3 {\n",
       "  background-color: #9dbdff;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_11054_row7_col4 {\n",
       "  background-color: #7396f5;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row8_col3 {\n",
       "  background-color: #536edd;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row8_col4 {\n",
       "  background-color: #7295f4;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_11054_row8_col5 {\n",
       "  background-color: #f39778;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_11054_row9_col3 {\n",
       "  background-color: #8badfd;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_11054_row9_col4 {\n",
       "  background-color: #c43032;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_11054\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_11054_level0_col0\" class=\"col_heading level0 col0\" >tissue</th>\n",
       "      <th id=\"T_11054_level0_col1\" class=\"col_heading level0 col1\" >cell_type</th>\n",
       "      <th id=\"T_11054_level0_col2\" class=\"col_heading level0 col2\" >disease</th>\n",
       "      <th id=\"T_11054_level0_col3\" class=\"col_heading level0 col3\" >control_prediction</th>\n",
       "      <th id=\"T_11054_level0_col4\" class=\"col_heading level0 col4\" >ko_prediction</th>\n",
       "      <th id=\"T_11054_level0_col5\" class=\"col_heading level0 col5\" >delta</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >cell_id</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "      <th class=\"blank col4\" >&nbsp;</th>\n",
       "      <th class=\"blank col5\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row0\" class=\"row_heading level0 row0\" >I175041_N1-ACCAGTACACTAGTAC</th>\n",
       "      <td id=\"T_11054_row0_col0\" class=\"data row0 col0\" >sigmoid colon</td>\n",
       "      <td id=\"T_11054_row0_col1\" class=\"data row0 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row0_col2\" class=\"data row0 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_11054_row0_col3\" class=\"data row0 col3\" >0.046</td>\n",
       "      <td id=\"T_11054_row0_col4\" class=\"data row0 col4\" >0.099</td>\n",
       "      <td id=\"T_11054_row0_col5\" class=\"data row0 col5\" >-0.053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row1\" class=\"row_heading level0 row1\" >H197396_N1-TTGGCAAAGATATGGT</th>\n",
       "      <td id=\"T_11054_row1_col0\" class=\"data row1 col0\" >colon</td>\n",
       "      <td id=\"T_11054_row1_col1\" class=\"data row1 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row1_col2\" class=\"data row1 col2\" >normal</td>\n",
       "      <td id=\"T_11054_row1_col3\" class=\"data row1 col3\" >0.660</td>\n",
       "      <td id=\"T_11054_row1_col4\" class=\"data row1 col4\" >0.899</td>\n",
       "      <td id=\"T_11054_row1_col5\" class=\"data row1 col5\" >-0.239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row2\" class=\"row_heading level0 row2\" >N20_Epi_B-AGATTCCTCCTTTA</th>\n",
       "      <td id=\"T_11054_row2_col0\" class=\"data row2 col0\" >colon</td>\n",
       "      <td id=\"T_11054_row2_col1\" class=\"data row2 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row2_col2\" class=\"data row2 col2\" >normal</td>\n",
       "      <td id=\"T_11054_row2_col3\" class=\"data row2 col3\" >0.919</td>\n",
       "      <td id=\"T_11054_row2_col4\" class=\"data row2 col4\" >0.949</td>\n",
       "      <td id=\"T_11054_row2_col5\" class=\"data row2 col5\" >-0.030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row3\" class=\"row_heading level0 row3\" >N128400_L-ACTACGATCGTTCTAT</th>\n",
       "      <td id=\"T_11054_row3_col0\" class=\"data row3 col0\" >lamina propria of mucosa of colon</td>\n",
       "      <td id=\"T_11054_row3_col1\" class=\"data row3 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row3_col2\" class=\"data row3 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_11054_row3_col3\" class=\"data row3 col3\" >0.122</td>\n",
       "      <td id=\"T_11054_row3_col4\" class=\"data row3 col4\" >0.478</td>\n",
       "      <td id=\"T_11054_row3_col5\" class=\"data row3 col5\" >-0.356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row4\" class=\"row_heading level0 row4\" >N178961_L-CTAACTTCACCCTAGG</th>\n",
       "      <td id=\"T_11054_row4_col0\" class=\"data row4 col0\" >lamina propria of mucosa of colon</td>\n",
       "      <td id=\"T_11054_row4_col1\" class=\"data row4 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row4_col2\" class=\"data row4 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_11054_row4_col3\" class=\"data row4 col3\" >0.041</td>\n",
       "      <td id=\"T_11054_row4_col4\" class=\"data row4 col4\" >0.025</td>\n",
       "      <td id=\"T_11054_row4_col5\" class=\"data row4 col5\" >0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row5\" class=\"row_heading level0 row5\" >H180844_N2-TGACAACTCACCCTCA</th>\n",
       "      <td id=\"T_11054_row5_col0\" class=\"data row5 col0\" >colon</td>\n",
       "      <td id=\"T_11054_row5_col1\" class=\"data row5 col1\" >epithelial cell</td>\n",
       "      <td id=\"T_11054_row5_col2\" class=\"data row5 col2\" >normal</td>\n",
       "      <td id=\"T_11054_row5_col3\" class=\"data row5 col3\" >0.022</td>\n",
       "      <td id=\"T_11054_row5_col4\" class=\"data row5 col4\" >0.003</td>\n",
       "      <td id=\"T_11054_row5_col5\" class=\"data row5 col5\" >0.020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row6\" class=\"row_heading level0 row6\" >N128624_E-TATTCCACAAGTAGTA</th>\n",
       "      <td id=\"T_11054_row6_col0\" class=\"data row6 col0\" >colonic epithelium</td>\n",
       "      <td id=\"T_11054_row6_col1\" class=\"data row6 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row6_col2\" class=\"data row6 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_11054_row6_col3\" class=\"data row6 col3\" >0.121</td>\n",
       "      <td id=\"T_11054_row6_col4\" class=\"data row6 col4\" >0.018</td>\n",
       "      <td id=\"T_11054_row6_col5\" class=\"data row6 col5\" >0.104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row7\" class=\"row_heading level0 row7\" >N130084_E-GGAGGTACAGGTTCCG</th>\n",
       "      <td id=\"T_11054_row7_col0\" class=\"data row7 col0\" >colonic epithelium</td>\n",
       "      <td id=\"T_11054_row7_col1\" class=\"data row7 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row7_col2\" class=\"data row7 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_11054_row7_col3\" class=\"data row7 col3\" >0.288</td>\n",
       "      <td id=\"T_11054_row7_col4\" class=\"data row7 col4\" >0.171</td>\n",
       "      <td id=\"T_11054_row7_col5\" class=\"data row7 col5\" >0.118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row8\" class=\"row_heading level0 row8\" >N51_Epi_A-CGGTTAAGTTCATGGT</th>\n",
       "      <td id=\"T_11054_row8_col0\" class=\"data row8 col0\" >caecum</td>\n",
       "      <td id=\"T_11054_row8_col1\" class=\"data row8 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row8_col2\" class=\"data row8 col2\" >normal</td>\n",
       "      <td id=\"T_11054_row8_col3\" class=\"data row8 col3\" >0.095</td>\n",
       "      <td id=\"T_11054_row8_col4\" class=\"data row8 col4\" >0.167</td>\n",
       "      <td id=\"T_11054_row8_col5\" class=\"data row8 col5\" >-0.073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_11054_level0_row9\" class=\"row_heading level0 row9\" >N175041_N2-CCTAGCTGTACCGGCT</th>\n",
       "      <td id=\"T_11054_row9_col0\" class=\"data row9 col0\" >left colon</td>\n",
       "      <td id=\"T_11054_row9_col1\" class=\"data row9 col1\" >enterocyte</td>\n",
       "      <td id=\"T_11054_row9_col2\" class=\"data row9 col2\" >Crohn disease</td>\n",
       "      <td id=\"T_11054_row9_col3\" class=\"data row9 col3\" >0.240</td>\n",
       "      <td id=\"T_11054_row9_col4\" class=\"data row9 col4\" >0.908</td>\n",
       "      <td id=\"T_11054_row9_col5\" class=\"data row9 col5\" >-0.667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7f210068fe50>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result[\"delta\"] = result[\"control_prediction\"] - result[\"ko_prediction\"]\n",
    "\n",
    "(\n",
    "    result.sample(10, random_state=23)\n",
    "    .style.format(\"{:.3f}\", subset=[\"control_prediction\", \"ko_prediction\", \"delta\"])\n",
    "    .background_gradient(\n",
    "        subset=[\"control_prediction\", \"ko_prediction\", \"delta\"], cmap=\"coolwarm\"\n",
    "    )\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44056"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# visualize the cells with JUND expression\n",
    "# find the cells with JUND expression:\n",
    "genes = adata.var.reset_index()[[\"feature_name\"]].copy()\n",
    "jund_index = genes[genes[\"feature_name\"] == \"JUND\"].index[0]\n",
    "jund_expression = adata.X[:, jund_index].toarray().flatten()\n",
    "# get the index of the cells with JUND expression:\n",
    "jund_cells = jund_expression > 0\n",
    "jund_cells_index = np.where(jund_cells)[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cross validation of a model, xgboost on the embeddings, we wantto predict the disease:\n",
    "# cross validation:\n",
    "from sklearn.model_selection import cross_val_score\n",
    "\n",
    "x = control_embedding[colon_index]\n",
    "y = adata.obs[\"disease\"].values[colon_index]\n",
    "y = pd.Categorical(y).codes\n",
    "model = xgboost.XGBClassifier()\n",
    "scores = cross_val_score(model, x, y, cv=5, scoring=\"accuracy\", n_jobs=5, verbose=1)\n",
    "\n",
    "print(f\"Accuracy: {scores.mean()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cross validation score of xgboost classifier trained to classify 'disease' state using the control embedding\n",
    "\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from xgboost import XGBClassifier\n",
    "\n",
    "x = control_embedding[colon_index]\n",
    "y = adata.obs[\"disease\"].iloc[colon_index]\n",
    "\n",
    "# y should be a categorical codes:\n",
    "y = pd.Categorical(y).codes\n",
    "\n",
    "scores = cross_val_score(XGBClassifier(), x, y, cv=5, scoring=\"accuracy\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "d4d1e4263499bec80672ea0156c357c1ee493ec2b1c70f0acce89fc37c4a6abe"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
